crates/extensions/tedge_mqtt_bridge/src/backoff.rs

```diff
@@ -0,0 +1,194 @@
+use std::ops::Deref;
+use std::time::Duration;
+use std::time::Instant;
+
+use ::backoff::backoff::Backoff;
+use ::backoff::exponential::ExponentialBackoff;
+use ::backoff::Clock;
+
+pub struct CustomBackoff<C> {
+    eb: ExponentialBackoff<C>,
+    last_failure: Option<Instant>,
+    reset_timeout: Duration,
+    last_state: State,
+}
+
+#[derive(PartialEq, Eq, Copy, Clone, Debug)]
+enum State {
+    Success,
+    Failure,
+}
+
+impl<C: Clock> CustomBackoff<C> {
+    pub fn new(
+        clock: C,
+        initial_interval: Duration,
+        max_interval: Duration,
+        reset_timeout: Duration,
+    ) -> Self {
+        Self {
+            eb: ExponentialBackoff {
+                current_interval: initial_interval,
+                initial_interval,
+                randomization_factor: 0.5,
+                multiplier: 2.0,
+                max_interval,
+                start_time: clock.now(),
+                max_elapsed_time: None,
+                clock,
+            },
+            last_failure: None,
+            reset_timeout,
+            last_state: State::Success,
+        }
+    }
+
+    #[must_use]
+    fn backoff(&mut self) -> BackoffHandle<C> {
+        let now = self.eb.clock.now();
+        match self.last_failure {
+            Some(time) if now - time < self.reset_timeout => (),
+            Some(_) | None if self.last_state == State::Success => {
+                // TODO check we actually succeeded on the previous try too
+                self.eb.reset()
+            }
+            _ => (),
+        };
+
+        self.last_state = State::Failure;
+
+        BackoffHandle {
+            duration: self.eb.next_backoff().unwrap(),
+            inner: self,
+        }
+    }
+
+    pub fn mark_success(&mut self) {
+        self.last_state = State::Success;
+    }
```

Comment:
There is something I don't understand.

I would expect this method to keep track of when occurred the first successful connection event, so we don't reset the timer if an error occurs before a minimum period with an effective connection. Note also, that I would track the time of the last success and not of the last failure.

```suggestion
    pub fn mark_success(&mut self) {
        if self.last_state == State::Failure {
            self.okay_since = now();
            self.last_state = State::Success;
        }
    }
```

What I'm missing?

Reply:
I think the answer is that you're missing nothing. I initially started by not relying on the consumer of this module manually "resetting" the backoff, but that's both needlessly complicated and also flawed (someone may reasonably want to disable backoff through setting `inital_interval` to 0).

Reply:
This commit implements what I have in mind https://github.com/thin-edge/thin-edge.io/pull/2915/commits/9e6ae7374a1e1a5f1cfc321ff0acd10f72a5ceef. Thanks